### ISSUE TITLE
`sort` and `sort_by` methods of Hash class

### ISSUE THREAD

**Issue by incetarik** [manastech/crystal#1915]
There is no sorting method in Hash class. How should we handle this? Or, will this be included in the future?

**Comment by jhass** [comment]
While `Hash` is insertion ordered, depending on that is usually a smell IMO. Are you sure you don't want an `Array({A, B})` actually?

**Comment by incetarik** [comment]
I would sort a hash by its values like `hash.sort_by { |_, v| v.size }` but I could not do as your said.

**Comment by jhass** [comment]
I didn't say that you can sort a `Hash`, I said that wanting to usually hints that you're using the wrong data structure.

**Comment by incetarik** [comment]
I understood. And, it would be better to use the structure you mentioned above. Thanks for help.

**Comment by rdp** [comment]
Hash#{sort, sort_by}  are convenience methods for .to_a.sort and might be nice (they exist in ruby-land anyway, but I suppose forcing the call to .to_a might clarify things hmm)

**Comment by oprypin** [comment]
For future reference,
to get a sorted array of pairs:

```crystal
hash.to_a.sort_by { |key, value| key }
```

**Comment by rmarronnier** [comment]
I don't want to revive an old corpse, but are there technical reasons why `Hash#sort_by` is not in the standard lib ?

@jhass : are there any reading materials I could work with to find alternate data structures to bypass this need ? Because I find myself using @oprypin one liners so many times...

**Comment by jhass** [comment]
I don't think there's much reading material because it's largely based on opinionated philosophy. So it's definitely not a technical reason.

Crystal often favors a little bit more explicitness in Ruby with the goal of reducing surprises. Here, I think it's easy to argue, that a method on `Hash` turns it into an `Array`, rather than changing the internal ordering a `Hash` has, can be a bit surprising offhand. In Ruby this happens because `Hash` is `Enumerable` and `sort_by` is an `Enumerable` method and the key thing to understand here is that in terms of `Enumerable` a `Hash` is a list of pairs and that's what `sort_by` operates on. But internalizing this connection, is quite some cognitive overhead to ask for.

So in Crystal so far we preferred to make this conversion a little bit more explicit, and the code a little bit more intention revealing. It literally says there, I want an ordered list of pairs, ordered by the following criteria.

Reiterating my previous point, personally I had very few cases where I ended up wanting to have the features of a map structure, that is constant time lookup of a value by some unique key, and the features of a list structure, that is a defined ordering between some values at the same time. In fact this feature combination is so seldom that many standard libraries don't offer any possibility for it, or at least do not provide it in the standard map time but rather a special one. Anyways, the few cases where I personally encountered this desire turned out to be quite indicative of me not having found the right datastructure for my problem yet, often simply realizing I didn't actually need one of the two features I talked about above and thus making the resulting program easier to reason about by using the other structure. This is what I meant when I said this is a code smell to me, it's an indication to me that I should be revisiting something.

Personally I wouldn't even use the term "one-liner" for the expression above, which I tend to use for common transformations on some data structure that usually have a much higher complexity than a call to `to_a` :) 

So, why does that little `to_a` bother you so much? :) Maybe your data should never be a `Hash` in the first place?

**Comment by rmarronnier** [comment]
Thanks for this extremely elaborate answer :-)
> Personally I wouldn't even use the term "one-liner" for the expression above

Haha ! Yes, you're right. Replacing one line of code with one another line, is not a oneliner :-p
Regarding the `sort_by` issue, IMO it's just synthetic sugar, mental load reduction and a gift to ruby newcomers.
It's not a big issue if this is not in the standard lib, but the cost of adding it is minimal.
I fell in love with Crystal because of its extreme readability. Thanks from borrowing from Ruby syntax and "encouraging" to declare types it makes reading code almost like reading natural language.
>  I said that wanting to usually hints that you're using the wrong data structure.
>  So, why does that little to_a bother you so much? :) Maybe your data should never be a Hash in the first place?

Ok, you have my full attention. I'm a junior developer, and using the key => value structure of Hash seems so obvious and useful to me for many tasks.
For example attributing a score value to an Array or Set of values then picking the 3 most scored ones, what would be the most clean/efficient way to do this without using Hash and a sort_by like method ?
[This is an old but recurrent example of using Hash for this problem](https://github.com/cadmiumcr/language_detector/blob/master/src/cadmium/language_detector/language_detector.cr)

**Comment by Fryguy** [comment]
I think my concern here is that `.to_a.sort_by` introduces a conversion to an Array of Tuples first, then calls sort_by on that intermediate Array.  A more direct sort_by method wouldn't need that intermediate Array and would instead work with .each directly.  I'd have to experiment to see if that is an _actual_ performance problem, but it does feel like, at a minimum, you'd have an extra copy of the collection in memory at once.

I think the confusion arises because sort_by in Crystal is only defined on Array (and Slice?), but in Ruby it's defined on Enumerable.  Was there a specific reason that sort_by was not defined on Enumerable in Crystal?

**Comment by jhass** [comment]
@Fryguy I just gave what I think the reason is above. Also I don't see the extra copy, sort algorithms need an array to operate, so when you chain `sort_by!` onto `to_a`, there's no extra. `Hash#sort_by`s implementation would very likely literally end up as `to_a.sort_by!`.

@rmarronnier 

> For example attributing a score value to an Array or Set of values then picking the 3 most scored ones, what would be the most clean/efficient way to do this without using Hash and a sort_by like method ?

It's generally hard to answer this in the abstract because it just depends so much on what you're doing. So take the following as responses to the example rather than me saying it's like this in general and there's no single exception :) 

So, in the example 

https://github.com/cadmiumcr/language_detector/blob/4020ac783e5962ff7910c17fe34d5c6d081b94cd/src/cadmium/language_detector/language_detector.cr#L34-L39

Here the return value could stay as `Array({String, Int32})`, the only usage of this is passing it to `get_distances` which just iterates it:

https://github.com/cadmiumcr/language_detector/blob/4020ac783e5962ff7910c17fe34d5c6d081b94cd/src/cadmium/language_detector/language_detector.cr#L84-L88

`get_distances` too has no reason to convert it back into a hash, its result is soley passed to 

https://github.com/cadmiumcr/language_detector/blob/4020ac783e5962ff7910c17fe34d5c6d081b94cd/src/cadmium/language_detector/language_detector.cr#L56-L62

which does not make any use of any hash features (`min = !distances.values[0..].empty? ? distances.values[0] : 1` -> `min = distances.dig(0, 1) || 1`).  The pattern continues, the only usage of the result of `normalize` is 

https://github.com/cadmiumcr/language_detector/blob/4020ac783e5962ff7910c17fe34d5c6d081b94cd/src/cadmium/language_detector/language_detector.cr#L43

which could just be `.dig(0, 0)` or `[0][0]`. 

In summary there's a lot of unneeded hashes in this example, getting rid of them would in my opinion make the program simpler (No question of "Why's this a hash?") and objectively faster due to less allocations.

If you want to regain some readability after, it probably makes sense to turn the pairs into a small struct like

```cr
record Score language : String, distance : Int32
```

That should have the same performance as a `Tuple` and suddenly you know what all these "keys" and "values" actually represent!

**Comment by rmarronnier** [comment]
Thanks @jhass for this fantastic and extensive answer !
I'll use #dig with Array(Tuple) now :-)
Also thanks for pointing out the `record` class : it's better than aliases or less overkill than a Struct.
> No question of "Why's this a hash?"
Now I realize, the reason I used hashes was because it works (and looks `some_hash[key]`) the same as a js object and the json output would be easier to work with on the frontend side of my web work.

**Comment by jhass** [comment]
> json output would be easier to work with on the frontend side of my web work.

That's a totally legit reason to convert it into a hash, I would just delay it as long as there's no benefit, in other words separate the presentation logic from the business logic :) Just keep in mind the [JSON spec](https://www.json.org/json-en.html) does not explicitly say that the order of keys in an JSON object is of semantic value, so that happening in any JSON encoder or decoder would be an implementation detail and something I wouldn't rely on when designing an JSON API :)

**Comment by rdp** [comment]
It might be interesting to create a "ruby compatibility" shard...or maybe one already exists?

**Comment by Blacksmoke16** [comment]
Going to bump this once again as i find myself wanting it every now and then. The main use cases I run into are using a hash to store types keyed by their "name".

For example `Route`s within a `RouteCollection`. The collection also has the concept of priorities which can be specified when a route is added. After all the routes are added and you want to iterate over them, they should be iterated in order based on their priority/insert order. Currently I'm needing to convert the hash to an array of tuples, do the sort, clear the hash, then re-insert everything back into it. I'm of course keeping track when the collection is in a sorted state and only need to this once (or when new routes are added), but still feels like there should be an easier way...

I'm also open to alternative implementations, but a hash seems to make the most sense so that the routes can be accessed by name versus needing to use like `Array#find`.

**Comment by rdp** [comment]
So maybe you're hoping for something like a TreeMap in java or something
like that?

**Comment by stellarpower** [comment]
Also I'm not sure on the implementation in Crystal, but in many hashing implementations (in e.g. other languages), it's a requirement that keys can be sorted/compared. set/map in the C++ stdlib, for example. I am using timestamps myself. Whilst it may indicate that I would be better using an array, in this case there is a definite order to my data. I tag values to the timestamp key, and when I want to output them, doing so in chronological order is what I need. Perhaps rather than sorting, an ordered "visitor" would be better - i.e. an equivalent of `each` or `map` that guarantees to yield in order of the key.

**Comment by rdp** [comment]
Do other languages have hashes that allow a "sort by" key for duplicate
entries?  In that case I guess they wouldn't return values by insertion
order but by this "other" order?

**Comment by straight-shoota** [comment]
@stellarpower I think this would be a use case for a different data structure with intrinsic ordering. Implementation of such a sorted map type would be more efficient based on a tree instead of a hash table.

To summarize, there are many different concepts of semantics for map types. And there are many possible implementations with specific tradeoffs for different use case.

**Comment by ysbaddaden** [comment]
Let's bump this again. I believe we can re-discuss the feature.

**`Hash` is ordered and that order is fixed to be the insertion order**. It's more than a de-facto standard, [the docs explicitly state](https://crystal-lang.org/api/1.19.1/Hash.html):

> Enumeration follows the order that the corresponding keys were inserted.

Currently, to sort a `Hash` we must instantiate a fresh `Hash` and add keys in the order we want, which usually requires an intermediate `Array` representation. For example:

```crystal
hsh2 = hsh1.to_a.sort!.to_h
```

I believe **there are no technical limitations** to sort a `Hash`. As stated above `Hash` is ordered and that's set in stone (won't change), to maintain the insertion order we need something to retain that order, and that something should be sortable.

Of course, it would only be sorted until we insert another key/value, but this is just as true for `Array`:

```crystal
ary.sort! { |item| item.size }      # <= sorted
ary << item                         # <= no longer sorted

hsh.sort! { |_, value| value.size } # <= sorted
hsh[key] = value                    # <= no longer sorted
```

A bonus would be the ability to skip the intermediate Array representation: `#sort!` could directly sort the index, and `#sort` be implemented as `dup.sort!`.

**The limitation is conceptual**. Other languages or libraries usually don't allow to sort a hashmap, even when the collection itself is sorted, it's always fixed to the insertion order:

- Ruby implements `Hash#sort` methods, but they return an array (`[[kv, ], ...]`);
- Python roughly does the same as current Crystal with `list2 = dict(sorted(list1))`;
- The `ordered_hash_map` Rust crate orders by insertion order (no manual sort);
- **C++ is an exception**: [`std::map`](https://en.cppreference.com/w/cpp/container/map.html) is ordered by key using a custom comparison (no arbitrary order of entries).

**Comment by straight-shoota** [comment]
Sounds good to me 👍 

It is already possible to reorder a hash with a bit of custom code. Even mutating the current instance instead of creating a duplicate is possible. And there are certainly use cases for this.
By providing a native implementation, it's much easier to use because it requires less boilerplate and we can probably sort more efficiently by using implementation internals.

This would not replace an implicitly sorted collection, of course.

**Comment by stellarpower** [comment]
As an alternative just to throw it out, what about something a bit like a mixin? As you say, there are always tradeoffs. It might be nice to have a generic class where the behaviour defaults to how it is now and 95% of cases won;t need to know or care, but if I want to change the "backend" but keep the same interface, then I can choose a different sort of hash, and then there needn't be several complicated methods on the one type. map/unordered_map might be a comparison here for this scenario - Hash/OrderedHash. I personally like the idea of mixing behaviours in à la template, in part because it allows combining options (I had this before for matrices - sparse/dense, immediate/lazy, etc.) but there is no need to implement it that way. A general map type where I can switch between hashing and trees, within some confines, a tree in general is a more flexible structure than a flat map, without breaking the interface feels like a nice hypothetical. Is there/should there be any parallel to Enumerable for mappings? I don't recall coming across one in Ruby. This also fits in with a sort of visitor/iterator/ranges idea where we should not couple too tightly on the storage. Sorting on the key could then in some cases be delegated to the iteration when we don't need to mutate the source. `to_a.sort.to_h` always irks me when I type it but it is convenient when I'm not worried about runtime performance but developer performance (typically the case for projects where I am using Crystal), maybe something like each_by where I could iterate with an arbitrary ordering and avoid trying to juggle keys.
